### PR TITLE
Shell openExternal can take optional callback (macOS)

### PR DESCRIPTION
**Pull request by gabriel**
As discussed in https://github.com/electron/electron/issues/6889

Added an optional callback argument to openExternal which will perform it asynchronously.

``` js
electron.shell.openExternal('http://electron.atom.io', {}, function(err) {
  if (err) console.log('Callback error:', err)
  else console.log('Callback opened')
})
```

I only implemented the asynchronous behavior on macOS.. I would need some help on the other platforms?

### CODE REVIEW

File: atom/common/platform_util_linux.cc [hunk 1]
```diff
@@ -88,6 +88,17 @@ bool OpenExternal(const GURL& url, bool activate) {
     return XDGOpen(url.spec(), false);
 }
 
+bool OpenExternal(const GURL& url, bool activate) {
```
Comment (enlight) [inline atom/common/platform_util_linux.cc]: This wrapper doesn't seem to serve any purpose right now.

File: atom/common/platform_util_win.cc [hunk 2]
```diff
@@ -316,6 +316,17 @@ bool OpenExternal(const base::string16& url, bool activate) {
   return true;
 }
 
+bool OpenExternal(const base::string16& url, bool activate) {
```
Comment (enlight) [inline atom/common/platform_util_win.cc]: This too is unnecessary right now.

File: docs/api/shell.md [hunk 3]
```diff
@@ -32,18 +32,22 @@ Returns `Boolean` - Whether the item was successfully opened.
...
-### `shell.openExternal(url[, options])`
+### `shell.openExternal(url[, options, callback])`
 
 * `url` String
 * `options` Object (optional) _macOS_
   * `activate` Boolean - `true` to bring the opened application to the
     foreground. The default is `true`.
+* `callback` Function (optional) _macOS_
 
 Returns `Boolean` - Whether an application was available to open the URL.
```
Comment (enlight) [inline docs/api/shell.md]: Is this still true if the call is async?

**Comment by gabriel** [comment]
To help review, this is what the docs now look like:

2016-10-14 at 12 43 pm

**Comment by sindresorhus** [comment]
IMHO, it should return a promise instead. All new async browser API's use promises.

**Comment by MarshallOfSound** [comment]
@sindresorhus I think for consistency our API's should stick with callbacks.

If at some point we want to use promises instead (I agree it's a much nicer / newer pattern) then we should move all API's across at once (possibly as part of 2.0 deprecation).  Having half and half would be quite confusing.

**Comment by paulcbetts** [comment]
Why is this useful to be asynchronous? Any OS where this call blocks for any amount of time is pretty broken imho

**Comment by sindresorhus** [comment]
@MarshallOfSound That means that for each new async API added until 2.0, there will be that many more breaking API's for users if 2.0 switches to promises. Consistency is nice, but nothing docs cannot solve, and future-proofing is more important.

**Comment by MarshallOfSound** [comment]
For those not interested in the promise discussion, you can skip over this

<details>
> That means that for each new async API added until 2.0, there will be that many more breaking API's for users if 2.0 switches to promises

We do not add new API's very regularly and I haven't seen an async one come through in a long time.  IMO that isn't too big of an issue.  Also I don't know if we could / should / will move to promises as it is one hell of a breaking change it was more of a "if we do it definitely shouldn't happen till then".

> Consistency is nice, but nothing docs cannot solve, and future-proofing is more important.

This is just IMO but I would normally take consistency over **partially** adopting new technologies.  Using a library that has a mixture of promises and callbacks requires constant doc lookups and adds extra hassle / time to what should be a simply experience.  Again, IMO, but I'd much rather know that either Electron is promise driven or callback driven.  Given that lots of Chromium's internals still use callbacks and wrapping those in promises would be time consuming and "temporary" code, I think we would be jumping on the Promise train is a bit too soon.

But this is a discussion for lots of people to have another time in another place if we want to discuss adopting promise logic.

*Argh I typed too much*
</details>

**Comment by paulcbetts** [comment]
@sindresorhus I'm agreeing with you In Principle, but I think before we can add promises in, we need a generalized way to marshal C++ code to return a Promise (i.e. right now I don't even know how I'd return a resolved promise via V8). Then we can retrofit all of these APIs to make the callback optional, similar to how many node libs are doing it so that both patterns are supported

**Comment by gabriel** [comment]
@paulcbetts It's nice to have an async call, because opening an email client or remote mount or document or whatever can take a long time (especially if accessing something remote)... having the ability to open and the get a callback with the status of the open is useful in these cases (showing in the UI, that it is opening).

The callback is optional and follows the same API as the [file dialog APIs](https://github.com/electron/electron/blob/master/docs/api/dialog.md) and is fully backward compatible.

**Comment by gabriel** [comment]
@enlight I think I covered all the requested changes, lmk if anything else needs tweaking, thx!

File: atom/common/platform_util_linux.cc [hunk 4]
```diff
@@ -88,6 +89,14 @@ bool OpenExternal(const GURL& url, bool activate) {
     return XDGOpen(url.spec(), false);
 }
 
+bool OpenExternal(const GURL& url, bool activate,
+  const OpenExternalCallback& callback) {
```
Comment (enlight) [inline atom/common/platform_util_linux.cc]: This line should either be aligned with the first function parameter above or indented by 4 spaces in total (whichever fits the ~80 column limit).

File: atom/common/platform_util_win.cc [hunk 5]
```diff
@@ -316,6 +316,14 @@ bool OpenExternal(const base::string16& url, bool activate) {
   return true;
 }
 
+bool OpenExternal(const base::string16& url, bool activate,
+  const OpenExternalCallback& callback) {
```
Comment (enlight) [inline atom/common/platform_util_win.cc]: Fix alignment.

File: atom/common/platform_util_win.cc [hunk 6]
```diff
@@ -316,6 +316,14 @@ bool OpenExternal(const base::string16& url, bool activate) {
...
+  // // TODO(gabriel): Implement async open if callback is specified
```
Comment (enlight) [inline atom/common/platform_util_win.cc]: Typo

File: atom/common/platform_util_mac.mm [hunk 7]
```diff
@@ -128,7 +129,17 @@ bool OpenItem(const base::FilePath& full_path) {
   return status == noErr;
 }
 
-bool OpenExternal(const GURL& url, bool activate) {
+bool openURLInWorkspace(NSURL* ns_url, NSUInteger launchOptions) {
```
Comment (zcbenz) [inline atom/common/platform_util_mac.mm]: We use AllCamelCase for function names, so `openURLInWorkspace` => `OpenURLInWorkspace`, `openExternal` => `OpenExternal`.

And helper functions should be put into anonymous namespace, you can find an example in `platform_util_linux.cc`.

Comment (gabriel) [inline atom/common/platform_util_mac.mm]: Moved helpers to anon namespace and fixed method names. Thanks!

**Comment by gabriel** [comment]
I think I covered all the requested changes, lmk if anything else needs tweaking, thx!

**Comment by YurySolovyov** [comment]
Would it be possible to keep node-style error-first callback signature to keep promisification simpler?
Error might signal that item is not found or something like that,

**Comment by gabriel** [comment]
Yeah, we probably should follow that pattern. I couldn't find any of the current callbacks in the  [API](https://github.com/electron/electron/tree/master/docs/api) passing an [Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) object as the first arg though, so not sure exactly how to do that, but I'll look into it.

**Comment by YurySolovyov** [comment]
@gabriel http://electron.atom.io/docs/api/web-contents/#contentsprinttopdfoptions-callback

**Comment by gabriel** [comment]
I updated to use an error for the callback. I was also able to simplify the code a little bit since LSGetApplicationForURL is also thread safe.

I'm able to determine the underlying error message in some cases from the `OSStatus` returned from LSGetApplicationForURL.

So for example, running these examples would return the following error messages:

```js
  electron.shell.openExternal('ooops', {}, function(err) {
    console.log('Should error with bad URL:', err)
  })

  electron.shell.openExternal('noappforscheme://oops', {}, function(err) {
    console.log('Should error with no application for URL:', err)
  })

  electron.shell.openExternal('file:///notfound', {}, function(err) {
    console.log('Should error with fail to open:', err)
  })
```

```
Should error with bad URL: Error: Invalid URL
    at Error (native)
    at App.<anonymous> (/Users/gabe/Projects/Electron/electron-quick-start/main.js:42:18)
    at emitTwo (events.js:111:20)
    at App.emit (events.js:191:7)

Should error with no application for URL: Error: No application in the Launch Services database matches the input criteria.
    at Error (native)

Should error with fail to open: Error: Failed to open (-43)
    at Error (native)
```

There is no error stack if the error happened asynchronously so I'm probably doing something wrong there?

This isn't very relevant but that last example, LSGetApplicationForURL returns -43 error code if file (via file://) isn't found, but I couldn't find it anywhere in CoreFoundation or LaunchServices source anywhere, so might be a Finder error code. (Not part of standard error codes via [NSError](http://nshipster.com/nserror/) or [LaunchServices](https://developer.apple.com/reference/coreservices/1658613-launch_services#1661359).)

**Comment by gabriel** [comment]
This might be ready to review again :)

**Comment by gabriel** [comment]
Let me know if there is anything I can do to help move this along.

This PR should be relatively safe since it's added behavior and code for the callback scenario and doesn't alter anything with the existing non-callback behavior.

**Comment by zcbenz** [comment]
👍

**Comment by zcbenz** [comment]
Sorry I messed up when trying to rebase this branch, it has been pushed to master even though this PR shows as closed.

**Comment by vivekdaramwalwy** [comment]
I need this asynchronous behavior for Windows as well. When I use openExternal() or openItem(), app hangs on Windows. Any solution?

**Comment by YurySolovyov** [comment]
Someone will probably need to implement one

**Comment by gabriel** [comment]
This was actually merged in even though shows as closed.

**Pull request closed without merging**